### DISCUSSION TITLE
Zephyr on Raspberry Pi 4 shows rainbow screen

### DISCUSSION THREAD

**Discussion by santosh-gawas** [zephyrproject-rtos/zephyr#73483] [Q&A]
Wanted to try out Zephyr for our new product. Started with Raspberry Pi 4. Followed the instructions on zephyr website as per the links below. It's not working not sure what I am doing wrong. Any help would be highly appreciated.

1. https://docs.zephyrproject.org/latest/develop/getting_started/index.html ====> zephyr.bin
2. https://docs.zephyrproject.org/latest/boards/raspberrypi/rpi_4b/doc/index.html ====> other files

1st attempt: Tried on Ubuntu VM on x86 system ---> Does not work.
2nd attempt: Tried on Raspberry Pi Ubuntu (aarch64) ---> Does not work.

Formatted the SD card with Raspberry Pi Imager/Rufus (FAT32).

**Comment by Jiayi0111** [comment]
**1st Attempt: Building on an Ubuntu VM (x86 System)**
If you can successfully build the Zephyr project on your Ubuntu VM and boot your Raspberry Pi 4, you should expect to see a rainbow screen if you are not running examples related to HDMI. This screen indicates that the firmware is loading but there is no HDMI output. Since the Raspberry Pi 4 does not have an onboard LED, you can try the "Hello World" example to verify if the system is working correctly.
To Connect via Serial:
Download PuTTY and connect it to your Raspberry Pi via serial.
Settings in PuTTY:
Connection type: Serial
Serial line: COM_ (replace with your actual COM port)
Speed: 115200
Go to Device Manager, find your COM port, and set the following:
Bits per second: 115200
Data bits: 8
Parity: None
Stop bits: 1
Flow control: None
Once everything is set up, you should be able to see the output in the PuTTY window 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/165753712/5f17f4a2-da10-41b0-94f0-f1dde01065e7)
**2nd Attempt: Building Directly on the Raspberry Pi**
I have not tried building Zephyr directly on a Raspberry Pi. The usual approach is to build on a more powerful x86 system and then transfer the build to the Raspberry Pi.

**Reply by santosh-gawas** [reply]
Ok I will try this later today once I get a serial cable. Can you please give me some idea to port and run cyclictest application on Zephyr RPI? 
https://wiki.linuxfoundation.org/realtime/documentation/howto/tools/rt-tests
Is there any RPI document to help enable HDMI and USB (overlay files)?

**Reply by Jiayi0111** [reply]
I am also trying to measure latency. What I did was generate a square wave on one GPIO pin and use another pin to trigger on the positive edge. By wiring these two pins together and capturing the CPU clock at the positive edge and at the point entering ISR, I measure the clk difference to determine latency.

However, I discovered that the Raspberry Pi 4 typically runs at a CPU speed of 600MHz instead of 1.5GHz. For faster response, CPU frequency scaling is needed. I've been struggling with this for three days. I believe it might be solved with an overlay, but I haven't been able to find any resources on the syntax or how to write an overlay for the Raspberry Pi 4B in Zephyr. If you have any resources or advice, please share them with me as well. Thank you!
